app/utils/IncomeSourcesUtils.scala

```diff
@@ -35,4 +41,27 @@ trait IncomeSourcesUtils extends FeatureSwitching {
       codeBlock
     }
   }
+  def withIncomeSourcesFSWithSessionCheck(sessionService: SessionService, journeyType: JourneyType)(codeBlock: => Future[Result])(implicit user: MtdItUser[_], hc: HeaderCarrier, ec: ExecutionContext): Future[Result] = {
+    if (isDisabled(IncomeSources)) {
+      user.userType match {
+        case Some(Agent) => Future.successful(Redirect(controllers.routes.HomeController.showAgent))
+        case _ => Future.successful(Redirect(controllers.routes.HomeController.show()))
+      }
+    } else {
+      journeyChecker(sessionService, journeyType).flatMap{
+        case true => user.userType match {
+          case Some(Agent) => Future.successful(Redirect(controllers.incomeSources.add.routes.YouCannotGoBackErrorController.showAgent(journeyType.businessType)))
+          case _ => Future.successful(Redirect(controllers.incomeSources.add.routes.YouCannotGoBackErrorController.show(journeyType.businessType)))
+        }
+        case false => codeBlock
+      }
+    }
+  }
+
+  def journeyChecker(sessionService: SessionService, journeyType: JourneyType)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Boolean] = {
```

Comment:
you suppressing error details in here, this is fine in case you really need this;
but I would consider to add error logging in so we can trace this even in Kibana

Reply:
For some reason the default false value in Mongo can't be retrieved so it usually gives an error even though things are acting as they should (the value has not been set to true yet). I can add the error logging in but I think it will be mostly false errors